extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -538,6 +567,167 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
+            endpoint, {credentials: 'include'}).then(res => {
+              rtcTotalTime = Date.now() - startTime;
+              /*
+              *  disableSWR should be set to true if the endpoint is not
+              *  returning cache headers.
+              */
+              verifyRtcConfigMember('disableStaleWhileRevalidate', 'boolean');
+              if (rtcConfig['disableStaleWhileRevalidate'] !== true) {
+                const headers = new Headers();
+                headers.append('Cache-Control', 'max-age=0');
+                // Repopulate the cache.
+                Services.xhrFor(this.win).fetchJson(endpoint, {
+                  credentials: 'include',
+                  headers,
+                }).catch(err => {
+                  user().error(TAG, err.message);
+                });
+              }
+              // Non-200 status codes are forbidden for RTC.
+              // TODO: Add to fetchResponse the ability to
+              // check for redirects as well.
+              if (res.status != 200) {
+                return {rtcTotalTime};
+              }
+              return res.text().then(text => {
+                // An empty text response is fine, just means
+                // we have nothing to merge.
+                if (!text) {
+                  return {rtcTotalTime, success: true};
+                }
+                const rtcResponse = tryParseJson(text);
+                return {rtcResponse, rtcTotalTime};
+              });
+            }));
+
+    return this.mergeRtc();
+  }
+
+  /**
+   * Merges the RTC response into the jsonTargeting of this.
+   * If it can't merge, or there is no response, potentially
+   * rejects.
+   * @return {Promise<?Object>} Resolves if ad request is
+   *     to be sent, with object of params to add to request,
+   *     otherwise rejects with a reject message if we have one.
+   */
+  mergeRtc() {
+    // add reasons for promise.reject
+    return rtcPromise.then(
+        r => {
+          // Don't try to merge if we're sending without RTC.
+          if (!r || (!r.rtcResponse && !r.success)) {
+            return this.shouldSendRequestWithoutRtc(
+                'Bad response');
+          } else if (!r.rtcResponse && r.success) {
+            // Empty response, no need to merge
+            return Promise.resolve({
+              artc: r.rtcTotalTime,
+              ati: RTC_ATI_ENUM.RTC_SUCCESS,
+              ard: parseUrl(rtcConfig['endpoint']).hostname,
+            });
+          }
+
+          const rtcResponse = r.rtcResponse;
+          const mergeHelper = key => {
+            if (!!rtcResponse[key]) {
+              this.jsonTargeting_[key] =
+                  !!this.jsonTargeting_[key] ?
+                  deepMerge(this.jsonTargeting_[key],
+                      rtcResponse[key]) :
+                                rtcResponse[key];
+            }
+          };
+          ['targeting', 'categoryExclusions'].forEach(key => {
```

Comment:
nit: can be forEach(key => mergeHelper) or move content of mergeHelper into forEach